File: 프로그래머스/3/43238. 입국심사/입국심사.js
const MAX = 1_000_000_000;

function solution(n, times) {
    return binarySearch(n, times);
}

function binarySearch(n, times) {
    const maxTime = Math.max(...times);
    let left = 1;
    let right = maxTime * n;
    while(left < right) {
        const mid = Math.floor((left + right) / 2)
        if (isPossible(n, mid, times)) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }
    return left;
}

function isPossible(n, limit, times) {
    let sum = 0;
    for(let time of times) {
        sum += Math.floor(limit / time);
    }
    return n <= sum;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    private static final long MAX = 1_000_000_000;
    
    public long solution(int n, int[] times) {
        return binarySearch(n, times);
    }
    
    private long binarySearch(int n, int[] times) {
        long left = 0;
        long right = MAX * MAX;
        while(left < right) {
            long mid = (left + right) / 2;
            if (isPossible(n, mid, times)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
    
    private boolean isPossible(int n, long limit, int[] times) {
        long sum = 0;
        for(int time : times) {
            sum += limit / time;
        }
        return sum >= n;
    }
}